Add IMEX for 6.12 NVIDIA R570 driver

yeazelm opened the pull request:
**Description of changes:**
This adds IMEX as a subpackage that provides support for GB200 devices which require the IMEX process to coordinate efficient communication between devices in ComputeDomains. The DRA device plugin will use these binaries when creating channels but GB200 hardware is required. This DRA DaemonSet provides [its own configuration file](https://github.com/NVIDIA/k8s-dra-driver-gpu/blob/4f10463592f4c877b0c252966d1b0ee14f3774a1/templates/compute-domain-daemon-config.tmpl.cfg) so only the binaries are provided. We do not provide a systemd service file either because the intention is to let the DRA driver manage this.

**Testing done:**
Built an `aws-k8s-1.33-nvidia` with this kit and confirmed the binaries run. 
```
bash-5.1# nvidia-imex --version
IMEX version is: 570.148.08
bash-5.1# nvidia-imex-ctl
Unable to open configuration file at /etc/nvidia-imex/config.cfg
Usage: nvidia-imex-ctl [-i <IP> <PORT>|-u <unix domain socket path>|-c <nvidia-imex config.cfg path>] <-n|-N|-q|-s|-d|-m> [-t <timeout>]
Where:
       -i <IP> <PORT>: Connect to the IMEX daemon's control service at the specified ip address and port.
       -u <unix domain socket path>: Connect to the IMEX daemon's control service at the specified unix domain socket path.
       -c <nvidia-imex config path>: Path to nvidia-imex configuration file.  Default value is /etc/nvidia-imex/config.cfg
NOTE:  If none of -i, -u, -c are provided, -c is assumed with default value.
       -n: Continuously monitor the entire IMEX domain status.  Requires config file (default or via -c)
       -N: Gets the full status of the entire IMEX domain.  Requires config file (default or via -c)
       -q: Query the status of the IMEX daemon once and return.
       -s: Subscribe to the status of the IMEX daemon.  Can be used with -t to terminate after a certain amount of time.
       -d: Dump performance metrics of the IMEX daemon once and return.
       -m: Subscribe to performance metrics of the IMEX daemon.  Can be used with -t to terminate after a certain amount of time.
       -t: Timeout (in milliseconds) for the -s command to terminate.  Unspecified means it will run until IMEX terminates.
       -r: Terminate status subscription when IMEX daemon returns READY status.
       -a: Only used with -q and -c.  Query status of all nodes configured, based on the config file in -c (or default.
       -j: Output of the -n/-N command will be in JSON, rather than formatted.
NOTE: only one of -i, -u, or -c can be specified.  If none are specified, -c is assumed with the default nvidia-imex config file location.
      also, only one command can be executed.
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
kmod-6.12-nvidia-r570: Add IMEX

This adds IMEX as a subpackage that provides support for GB200 devices
which require the IMEX process to coordinate efficient communication
between devices in ComputeDomains.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

arnaldo2792 commented on packages/kmod-6.12-nvidia-r570/Cargo.toml:
```diff
@@ -36,6 +36,16 @@ url = "https://developer.download.nvidia.com/compute/cuda/repos/rhel9/sbsa/nvidi
 force-upstream = true
 
+[[package.metadata.build-package.external-files]]
+url = "https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/nvidia-imex-570-570.148.08-1.x86_64.rpm"
```
I'll create us a task to migrate all of these to:

https://developer.download.nvidia.com/compute/cuda/repos/amzn2023/x86_64/

arnaldo2792 replied:
I'll pause on this ticket for now.

arnaldo2792 commented on packages/kmod-6.12-nvidia-r570/kmod-6.12-nvidia-r570.spec:
```diff
@@ -141,6 +153,11 @@ rpm2cpio %{_sourcedir}/nvidia-fabric-manager-%{tesla_ver}-1.%{_cross_arch}.rpm |
 # Add the license.
 install -p -m 0644 %{S:2} %{S:3} .
 
+# Extract fabricmanager from the rpm via cpio rather than `%%setup` since the
```
Copy/paste issue: `s/fabricmanager/imex`.

yeazelm replied:
Nice catch, I'll fix.

arnaldo2792 commented on packages/kmod-6.12-nvidia-r570/kmod-6.12-nvidia-r570.spec:
```diff
@@ -141,6 +153,11 @@ rpm2cpio %{_sourcedir}/nvidia-fabric-manager-%{tesla_ver}-1.%{_cross_arch}.rpm |
 
+# Extract fabricmanager from the rpm via cpio rather than `%%setup` since the
+# correct source is architecture-dependent.
+mkdir imex-%{fm_arch}-%{tesla_ver}-archive
```
Since `fm_arch` now is used for a different purpose, I think we should rename it to something else. Probably as the first commit in this PR so that the second commit only includes the actual changes of interest.

yeazelm replied:
I can do that, it felt strange to me as well and we are really stretching it now.

arnaldo2792 commented on packages/kmod-6.12-nvidia-r570/kmod-6.12-nvidia-r570.spec:
```diff
@@ -708,3 +732,8 @@ popd
 %{_cross_unitdir}/nvidia-fabricmanager.service
+
+%files imex
+%license imex-%{fm_arch}-%{tesla_ver}-archive/usr/share/licenses/nvidia-imex-%{tesla_major}/LICENSE
```
If we are doing this now, we should also update the fabric manager as its RPM also provides a license:

```bash
rpm -q --list nvidia-fabric-manager-570.148.08-1.x86_64.rpm | rg LICENSE
/usr/share/doc/nvidia-fabricmanager/LICENSE
```

arnaldo2792 commented on packages/kmod-6.12-nvidia-r570/kmod-6.12-nvidia-r570.spec:
```diff
@@ -85,10 +89,18 @@ Requires: %{name}-grid
 Summary: NVIDIA fabricmanager config and service files
 Requires: %{name}-tesla(fabricmanager)
 Requires: %{_cross_os}nvlsm
+Requires: %{name}-imex
```
Are you adding `imex` here because it is a dependency of `fabricmanager` and therefore a dependency of the `tesla` driver?

yeazelm replied:
Correct, Fabric Manager is now responsible for initialization of many of these chips like the B200, and IMEX is only relevant on GB200 at the moment. So the rationale is to keep the dependencies all at the fabric manager level. This is all pretty intertwined at this point but this felt like the right way to have it naturally included.

bcressey commented on packages/kmod-6.12-nvidia-r570/kmod-6.12-nvidia-r570.spec:
```diff
@@ -708,3 +732,8 @@ popd
 %{_cross_unitdir}/nvidia-fabricmanager.service
+
+%files imex
+%license imex-%{fm_arch}-%{tesla_ver}-archive/usr/share/licenses/nvidia-imex-%{tesla_major}/LICENSE
```
Is the license different from the overall package license, such that the fm and imex subpackages should have a different license statement?

yeazelm replied:
The Fabric Manager and IMEX licenses are the same. I can drop this, I was just over zealous with tracking it.

Commit:
kmod-6.12-nvidia-r570: Change to generic arch variable

The variable for matching NVIDIA's upstream naming of architecture is
becoming more useful than just for Fabric Manager. This changes the
variable to a more generic name that expands beyond Fabric Manager.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
kmod-6.1-nvidia-r570: Change to generic arch variable

The variable for matching NVIDIA's upstream naming of architecture is
becoming more useful than just for Fabric Manager. This changes the
variable to a more generic name that expands beyond Fabric Manager.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
nvlsm: Change to generic arch variable

The variable for matching NVIDIA's upstream naming of architecture is
becoming more useful than just for Fabric Manager. This changes the
variable to a more generic name that expands beyond Fabric Manager.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
kmod-5.15-nvidia: Change to generic arch variable

The variable for matching NVIDIA's upstream naming of architecture is
becoming more useful than just for Fabric Manager. This changes the
variable to a more generic name that expands beyond Fabric Manager.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
kmod-6.1-nvidia: Change to generic arch variable

The variable for matching NVIDIA's upstream naming of architecture is
becoming more useful than just for Fabric Manager. This changes the
variable to a more generic name that expands beyond Fabric Manager.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

yeazelm commented:
^ Adds in the rename for `fm_arch` to `nvidia_arch` in separate commits along with addressing comments from the PR.

bcressey commented on packages/kmod-6.12-nvidia-r570/kmod-6.12-nvidia-r570.spec:
```diff
@@ -708,3 +732,8 @@ popd
 %{_cross_unitdir}/nvidia-fabricmanager.service
+
+%files imex
+
```
remove newline after `%files imex`:
```suggestion
```

yeazelm replied:
Fixed in recent push.
